### PR TITLE
Add ParticleNumberMeasurement support to the fermionic GaussianSimulator

### PR DESCRIPTION
**Pull request by simsaidan**
This PR enables particle-number sampling for pq.fermionic.GaussianSimulator. Computing all 2^n output probabilities is avoided by using mode-by-mode (chain-rule) sampling, adapting the Gaussian-Boson-Sampling strategy to fermions. For each prefix we reduce the state to the relevant modes and read off the joint occupation probability via GaussianState.get_particle_detection_probability.

Addresses #521 

Planning the implementation was done using Cursor

### CODE REVIEW

File: piquasso/fermionic/gaussian/simulation_steps.py [hunk 1]
```diff
@@ -273,3 +277,92 @@ def ising_XX(state: GaussianState, instruction: IsingXX, shots: int) -> "List[Br
...
+    modes = instruction.modes
+    fallback_np = state._connector.fallback_np
+
+    @lru_cache(state._config.cache_size)
+    def get_probability(
+        subspace_modes: "Tuple[int, ...]", occupation_numbers: "Tuple[int, ...]"
+    ) -> float:
+        reduced_state = state.reduced(subspace_modes)
+
+        with fallback_np.errstate(invalid="ignore"):
```
Comment (Kolarovszki) [inline piquasso/fermionic/gaussian/simulation_steps.py]: This is a bit alarming: why do you need this? I don't think that we should silently ignore such warnings.

File: piquasso/fermionic/gaussian/simulation_steps.py [hunk 2]
```diff
@@ -273,3 +277,92 @@ def ising_XX(state: GaussianState, instruction: IsingXX, shots: int) -> "List[Br
...
+                    reduced_state.get_particle_detection_probability(
+                        fallback_np.array(occupation_numbers)
+                    )
+                )
+            )
+
+        # `get_particle_detection_probability` may return `nan` (or a tiny negative
+        # value) when the true probability is essentially zero, since the determinant
+        # under the square root can dip slightly below zero numerically.
+        if fallback_np.isnan(probability) or probability < 0.0:
```
Comment (Kolarovszki) [inline piquasso/fermionic/gaussian/simulation_steps.py]: Again, I don't agree that we should handle nan values like this. We should handle nans immediately at the source, if possible.

File: tests/fermionic/gaussian/test_measurements.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+    with np.errstate(invalid="ignore"):
+        exact_probabilities = np.nan_to_num(state.fock_probabilities, nan=0.0)
+
+    with pq.Program() as program:
+        pq.Q() | pq.StateVector([1, 0, 1, 0])
+        pq.Q(all) | pq.Interferometer(U)
+        pq.Q() | pq.ParticleNumberMeasurement()
+
+    shots = 20000
```
Comment (Kolarovszki) [inline tests/fermionic/gaussian/test_measurements.py]: I think this many samples are a bit excessive in a unit test. Moreover, there is a slight chance that such test fails, and I fear that it is flaky. Also, these kinds of tests are a bit hard to debug sometimes.

**Review by Kolarovszki — CHANGES_REQUESTED** [review]
Overall, I think the approach is good, but I have some questions/comments.

File: tests/fermionic/gaussian/test_measurements.py (hunk 3 above)
Comment (simsaidan) [reply tests/fermionic/gaussian/test_measurements.py]: @Kolarovszki what would be a better value here? like 1000?

**Comment by simsaidan** [comment]
@Kolarovszki I think the nans were just from floating point issues. I made changes to make handling more natural. let me know if what I have now is better!

**Comment by simsaidan** [comment]
Hey @Kolarovszki ! Hope you are doing well. Just wanted to give this PR a quick bump in case it slipped through the cracks. Thanks!

**Review by Kolarovszki — APPROVED** [review]
Sorry for the late feedback. Looks good!

**Comment by simsaidan** [comment]
Thanks @Kolarovszki ! Could you assign the issue to me for tracking?

File: tests/fermionic/gaussian/test_measurements.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    state = state_simulator.execute(state_program).state
+
+    exact_mean = state.mean_particle_numbers(modes=tuple(range(d)))
+
+    with pq.Program() as program:
+        pq.Q() | pq.StateVector([1, 1, 0, 1, 0])
+        pq.Q(all) | pq.Interferometer(U)
+        pq.Q() | pq.ParticleNumberMeasurement()
+
+    shots = 1000
```
Comment (Kolarovszki) [inline tests/fermionic/gaussian/test_measurements.py]: Oh, sorry, I just noticed this again. As we discussed, such tests are not really good. Look at similar tests in the `tests/` folder testing `ParticleNumberMeasurement`, they don't test like this. Such tests can easily be flaky. If you convince me that there is an astronomically small chance that the assertion fails, I accept this, but otherwise, I wouldn't merge it like this.

Tests that are not flaky:
- Testing that we get the same samples regardless of connector class.
- Testing that the same seed reproduces the samples.
- Edge cases, like identity, single beamsplitter, blocked interferometer matrix, permutation matrix, do the output make sense in these cases?
- vacuum is mapped to vacuum.
- particle number conservation.
- etc.

But testing that the particle number detection reproduces the exact distribution is neither efficient nor stable.

**Review by Kolarovszki — CHANGES_REQUESTED** [review]
I still have one comment regarding the tests, otherwise, the production code looks good!

**Comment by simsaidan** [comment]
@Kolarovszki That makes sense. I added a new commit that uses less flaky tests. Is the new test file better?

**Review by Kolarovszki — APPROVED** [review]
Looks good to me! Thank you for your contribution!

**Comment by simsaidan** [comment]
@Kolarovszki For some reason, Ci was failing due to one of the tutorials. I made a 1 line change and now it is green.

**Comment by Kolarovszki** [comment]
@simsaidan Oh, turning off dask is not the solution. Let me investigate this issue -- I will assign the issue to you.

**Comment by Kolarovszki** [comment]
Can you help me finding the failing job? I have found one, but it fails due to a copyright error instead: https://github.com/Budapest-Quantum-Computing-Group/piquasso/actions/runs/27322278403

**Comment by Kolarovszki** [comment]
@simsaidan OK, so I think this is flaky. Can you undo editing the tutorial, and try to run the jobs again? It should pass, eventually. Sorry for the inconvenience.

**Comment by simsaidan** [comment]
@Kolarovszki Undid the tutorial edit and now CI is green!

**Comment by simsaidan** [comment]
@Kolarovszki Sorry to bump you again, but could we close the github issue since this is merged?
